lib/spack/spack/package.py

```diff
@@ -673,17 +681,40 @@ def url_for_version(self, version):
             version = Version(version)
 
         cls = self.__class__
-        if not (hasattr(cls, 'url') or self.version_urls()):
-            raise NoURLError(cls)
-
-        # If we have a specific URL for this version, don't extrapolate.
-        version_urls = self.version_urls()
-        if version in version_urls:
-            return version_urls[version]
 
-        # If we have no idea, try to substitute the version.
-        return spack.url.substitute_version(
-            self.nearest_url(version), self.url_version(version))
+        if hasattr(cls, 'url') or self.version_urls():
+            # If we have a specific URL for this version, don't extrapolate.
+            version_urls = self.version_urls()
+            if version in version_urls:
+                return version_urls[version]
+
+            # If we have no idea, try to substitute the version.
+            return spack.url.substitute_version(
+                self.nearest_url(version), self.url_version(version))
+        elif hasattr(cls, 'pypi') or hasattr(self, 'pypi'):
+            client = xmlrpclib.ServerProxy('https://pypi.python.org/pypi')
+            releases = client.release_urls(cls.pypi, str(version))
+
+            # Each release may contain dozens of different download URLs
+            # for wheels, tarballs, or zipballs. We don't really have a good
+            # way to tell which version the package needs to download.
+            # Tarballs have the highest preference
+            for release in releases:
+                if release['filename'].endswith('.tar.gz'):
+                    return release['url']
+
+            # Zipballs have a lower preference
+            for release in releases:
+                if release['filename'].endswith('.zip'):
+                    return release['url']
+
+            # Are there any source distributions at all?
+            for release in releases:
+                if release['packagetype']    == 'sdist' or \
+                   release['python_version'] == 'source':
+                    return release['url']
```

Comment:
The FetchStrategy stuff is very opaque to me.  I don't know when or how Spack decides which FetchStrategy ot instantiate.  Maybe that's why.